### PR TITLE
Support retry after flush

### PR DESCRIPTION
**Pull request by viglucci**
### Pull Request check-list

_Please make sure to review and check all of these items:_

- [x] Does `npm test` pass with this change (including linting)?
- [x] Is the new or changed code fully tested?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)?

### Description of change

Allows for the retry_strategy to return a object which results in the command queue being flushed but also for a connection retry attempt to occur. This change is in response to #1198  .

### CODE REVIEW

**Comment by viglucci** [comment]
I think there is some edge case with the AppVeyor tests that cause a slight miss timing and cause one of the tests to fail, I same a similar or the same error in a few of the other pull requests that are also marked as not successful.

File: test/connection.spec.js [hunk 1]
```diff
@@ -322,8 +342,8 @@ describe('connection tests', function () {
                 });
 
                 it('retryStrategy used to reconnect with defaults', function (done) {
-                    var unhookIntercept = intercept(function () {
-                        return '';
+                    var unhookIntercept = intercept(function (text) {
+                        return "";
```
Comment (BridgeAR) [inline test/connection.spec.js]: This should be kept as it was before.

File: test/connection.spec.js [hunk 2]
```diff
@@ -303,6 +303,26 @@ describe('connection tests', function () {
...
+                                client.set('foo', 'bar', function (err, res) {
+                                    assert.strictEqual(err.message, 'Stream connection ended and command aborted.');
+                                    assert.strictEqual(err.code, 'NR_CLOSED');
+                                    assert.strictEqual(err.origin.code, 'ECONNREFUSED');
+                                    done();
+                                });
+                            }
+                            return {
+                                retry_delay: Math.min(options.attempt * 25, 200),
+                                error: options.error
```
Comment (BridgeAR) [inline test/connection.spec.js]: It would be better to return a individual error. That way we'd know that the calling function indeed received that error.

File: index.js [hunk 3]
```diff
@@ -616,14 +616,25 @@ RedisClient.prototype.connection_gone = function (why, error) {
             if (this.retry_delay instanceof Error) {
                 error = this.retry_delay;
             }
+            // Handle object pattern
+            if (typeof this.retry_delay === 'object' && !(this.retry_delay instanceof Error) && this.retry_delay !== null) {
```
Comment (BridgeAR) [inline index.js]: The `&& !(this.retry_delay instanceof Error)` is a redundant check. Please just use a `else if` instead.

File: test/connection.spec.js (hunk 2 above)
Comment (viglucci) [inline test/connection.spec.js]: I'm not sure I follow. Do you mean just in the context of this test case, or in the implementation? With this change we have to return the error and the retry_delay, otherwise the reconnect is never attempted.

**Comment by viglucci** [comment]
I adressed your comments above @BridgeAR ,  however that seems to have upset the CI :(.
